File: Receiver_RX.java
import java.io.FileOutputStream;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.security.MessageDigest;
import java.util.HashMap;
import java.util.Map;

public class Receiver_RX {
    private static final int PORT = 5005;
    private static final int PUF_SIZE = 65536;

    public static void main(String[] args) throws Exception {
        DatagramSocket socket = new DatagramSocket(PORT);
        Map<Integer, byte[]> speicher = new HashMap<>();
        Integer erwarteteId = null, maxSeq = null;
        String dateiname = null;
        byte[] erwartetesMd5 = null;

        System.out.println("Receiver bereit...");

        while (true) {
            byte[] buf = new byte[PUF_SIZE];
            DatagramPacket pkt = new DatagramPacket(buf, buf.length);
            socket.receive(pkt);
            byte[] data = pkt.getData();
            int len = pkt.getLength();

            // TX_ID
            int txId = ((data[0]&0xFF)<<8) | (data[1]&0xFF);
            // Seq
            int seq  = ((data[2]&0xFF)<<24)|((data[3]&0xFF)<<16)|((data[4]&0xFF)<<8)|(data[5]&0xFF);

            // Erstes Paket
            if (seq == 0) {
                erwarteteId = txId;
                maxSeq = ((data[6]&0xFF)<<24)|((data[7]&0xFF)<<16)|((data[8]&0xFF)<<8)|(data[9]&0xFF);
                int nameLen = ((data[10]&0xFF)<<24)|((data[11]&0xFF)<<16)|((data[12]&0xFF)<<8)|(data[13]&0xFF);
                dateiname = new String(data, 14, nameLen, "UTF-8");
                System.out.printf("[RX] Empfang '%s', erwarte %d Pakete%n", dateiname, maxSeq);
            }
            // Letztes Paket
            else if (maxSeq != null && seq == maxSeq - 1) {
                erwartetesMd5 = new byte[16];
                System.arraycopy(data, 6, erwartetesMd5, 0, 16);
                System.out.println("[RX] Letztes Paket empfangen.");
            }
            // Datenpaket
            else if (txId == erwarteteId) {
                int payloadLen = len - 6;
                byte[] chunk = new byte[payloadLen];
                System.arraycopy(data, 6, chunk, 0, payloadLen);
                speicher.put(seq, chunk);
                System.out.printf("[RX] Paket %d (%d B) zwischengespeichert%n", seq, payloadLen);
            }

            // Fertig, wenn alle Datenpakete + Footer da sind
            if (erwartetesMd5 != null && speicher.size() == maxSeq - 2) {
                // Datei schreiben & MD5 prüfen
                MessageDigest md5 = MessageDigest.getInstance("MD5");
                try (FileOutputStream fos = new FileOutputStream("recv_" + dateiname)) {
                    for (int s = 1; s < maxSeq - 1; s++) {
                        byte[] chunk = speicher.get(s);
                        fos.write(chunk);
                        md5.update(chunk);
                    }
                }
                byte[] actual = md5.digest();
                if (MessageDigest.isEqual(actual, erwartetesMd5)) {
                    System.out.println("[RX] Datei korrekt empfangen (MD5 OK).");
                } else {
                    System.out.println("[RX] MD5 mismatch – Übertragung fehlerhaft.");
                }
                break;
            }
        }
        socket.close();
    }
}

File: Receiver_RX.py
import socket
import hashlib
import os

PORT = 5005
BUFFER_SIZE = 65536

def main():
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.bind(('', PORT))
    packets = {}
    expected_txid = None
    max_seq = None
    filename = None
    expected_md5 = None

    print("Receiver ready...")
    start = None

    while True:
        data, addr = sock.recvfrom(BUFFER_SIZE)
        if start is None:
            start = os.times()

        if len(data) < 6:
            continue
        txid = int.from_bytes(data[0:2], 'big')
        seq  = int.from_bytes(data[2:6], 'big')

        # First Packet
        if seq == 0:
            expected_txid = txid
            max_seq = int.from_bytes(data[6:10], 'big')
            name_len = int.from_bytes(data[10:14], 'big')
            filename = data[14:14+name_len].decode()
            print(f"Receiving '{filename}', maxSeq={max_seq}")

        # Last Packet
        elif seq == max_seq-1:
            expected_md5 = data[6:22]
            print(f"Last packet received (seq {seq})")

        # Data-Packet
        else:
            if txid == expected_txid:
                packets[seq] = data[6:]
                print(f"Packet {seq} stored ({len(data)-6} B)")

                # ACK send
                ack = txid.to_bytes(2, 'big') + seq.to_bytes(4, 'big')
                sock.sendto(ack, addr)

        # Check complete
        if expected_md5 and len(packets) == max_seq-2:
            # reconstruct
            hasher = hashlib.md5()
            with open(f"recv_{filename}", 'wb') as f:
                for s in range(1, max_seq-1):
                    chunk = packets[s]
                    f.write(chunk)
                    hasher.update(chunk)
            comp = hasher.digest()
            if comp == expected_md5:
                print("File received correctly (MD5 ok).")
            else:
                print("MD5 mismatch — Übertragung fehlerhaft.")
            break

    sock.close()

if __name__ == "__main__":
    main()
